File: ejemplos/nerve/05-logger-distribuido/ejemplo.js
/**
 * Ejemplo: Escritura de Logs en Node.js
 * 
 * Qué enseña este ejemplo:
 *   - Cómo obtener la fecha y hora actual usando el objeto nativo `Date`.
 *   - Cómo formatear una fecha al estilo ISO o local.
 *   - Cómo añadir texto al final de un archivo usando `fs.appendFileSync` (modo append).
 * 
 * Para tu reto (Logger Distribuido con Nerve):
 *   En vez de llamar a `escribirLog` con datos estáticos, tu receptor de Nerve
 *   escuchará mensajes de otros nodos y llamará a tu función de log para 
 *   registrarlos en un archivo central.
 * 
 * Glosario:
 *   new Date()         — Crea un nuevo objeto de fecha con la fecha y hora actual.
 *   toISOString()      — Formatea la fecha al estándar ISO (ej: 2026-07-17T15:30:05.123Z).
 *   fs.appendFileSync  — Escribe en un archivo síncronamente agregando el texto al final. 
 *                        Si el archivo no existe, lo crea.
 */

const fs = require('fs');
const path = require('path');

// El archivo donde guardaremos los logs
const ARCHIVO_LOG = path.join(__dirname, 'sistema.log');

function escribirLog(nivel, mensaje) {
    // 1. Obtenemos el timestamp actual y lo formateamos
    // toISOString da un formato como "2026-07-17T15:30:05.123Z"
    // Reemplazamos la "T" por un espacio y quitamos los milisegundos y la Z final para que luzca mejor.
    const ahora = new Date();
    const timestampFormateado = ahora.toISOString().replace('T', ' ').substring(0, 19);

    // 2. Construimos la línea de log final
    // Ejemplo: [2026-07-17 15:30:05] [INFO] El servidor ha iniciado correctamente
    const lineaLog = `[${timestampFormateado}] [${nivel}] ${mensaje}\n`;

    // 3. Imprimimos en consola (opcional, para verlo en vivo)
    // Usamos process.stdout.write para evitar el salto de línea extra que pone console.log
    process.stdout.write(lineaLog);

    // 4. Escribimos en el archivo agregando al final (append)
    try {
        fs.appendFileSync(ARCHIVO_LOG, lineaLog, 'utf8');
    } catch (error) {
        console.error('Error al escribir en el archivo de log:', error);
    }
}

function main() {
    console.log(`Escribiendo logs en '${ARCHIVO_LOG}'...\n`);

    escribirLog("INFO", "Sistema iniciado.");
    escribirLog("DEBUG", "Conexión a la base de datos establecida en 14ms.");
    escribirLog("WARNING", "Poco espacio en disco (15% restante).");
    escribirLog("ERROR", "Fallo al procesar el archivo 'datos.csv': Archivo no encontrado.");
    escribirLog("INFO", "Apagando sistema de forma segura.");

    console.log(`\nRevisa el archivo '${ARCHIVO_LOG}' para ver los resultados guardados.`);
}

main();

File: archivo-historico/ejemplos/nerve/05-logger-distribuido/ejemplo.go
/*
 * Ejemplo: Escritura de Logs en Go
 * 
 * Qué enseña este ejemplo:
 *   - Cómo obtener la fecha y hora actual usando `time.Now()`.
 *   - Cómo formatear una fecha en Go usando su sistema de fechas de referencia.
 *   - Cómo abrir un archivo en modo Append usando `os.OpenFile` y los flags correctos.
 * 
 * Para tu reto (Logger Distribuido con Nerve):
 *   En vez de llamar a `escribirLog` con datos estáticos, tu receptor de Nerve
 *   escuchará mensajes de otros nodos y llamará a tu función de log para 
 *   registrarlos en un archivo central.
 * 
 * Glosario:
 *   time.Now()          — Obtiene la fecha y hora actual.
 *   Format("2006-01-02 15:04:05") — Formatea la fecha. OJO: Go usa una fecha de referencia exacta
 *                         (2 de enero de 2006 a las 15:04:05) para definir formatos, NO usa %Y-%m-%d.
 *   os.OpenFile         — Abre un archivo con opciones avanzadas.
 *   os.O_APPEND         — Flag: si escribes, hazlo al final del archivo.
 *   os.O_CREATE         — Flag: si el archivo no existe, créalo.
 *   os.O_WRONLY         — Flag: abre el archivo solo para escritura.
 *   f.WriteString       — Escribe un string en el archivo abierto.
 */

package main

import (
	"fmt"
	"log"
	"os"
	"time"
)

const ArchivoLog = "sistema.log"

func escribirLog(nivel string, mensaje string) {
	// 1. Obtenemos el timestamp actual y lo formateamos
	// Recordatorio: "2006-01-02 15:04:05" es la forma de Go para decir "YYYY-MM-DD HH:MM:SS"
	ahora := time.Now()
	timestampFormateado := ahora.Format("2006-01-02 15:04:05")

	// 2. Construimos la línea de log final
	// Ejemplo: [2026-07-17 15:30:05] [INFO] El servidor ha iniciado correctamente
	lineaLog := fmt.Sprintf("[%s] [%s] %s\n", timestampFormateado, nivel, mensaje)

	// 3. Imprimimos en consola (opcional, para verlo en vivo)
	fmt.Print(lineaLog)

	// 4. Abrimos el archivo en modo "Append" (añadir al final)
	// Si no existe (O_CREATE), lo crea. Solo escritura (O_WRONLY).
	// Permisos 0644 (lectura y escritura para el dueño, solo lectura para otros).
	f, err := os.OpenFile(ArchivoLog, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
	if err != nil {
		log.Printf("Error abriendo el archivo de log: %v", err)
		return
	}
	defer f.Close()

	// 5. Escribimos la línea en el archivo
	if _, err := f.WriteString(lineaLog); err != nil {
		log.Printf("Error escribiendo en el archivo de log: %v", err)
	}
}

func main() {
	fmt.Printf("Escribiendo logs en '%s'...\n\n", ArchivoLog)

	escribirLog("INFO", "Sistema iniciado.")
	escribirLog("DEBUG", "Conexión a la base de datos establecida en 14ms.")
	escribirLog("WARNING", "Poco espacio en disco (15% restante).")
	escribirLog("ERROR", "Fallo al procesar el archivo 'datos.csv': Archivo no encontrado.")
	escribirLog("INFO", "Apagando sistema de forma segura.")

	fmt.Printf("\nRevisa el archivo '%s' para ver los resultados guardados.\n", ArchivoLog)
}
